TestNewInflightLimitMiddleware test flakes

The `TestNewInflightLimitMiddleware` test appears to fail pretty consistently (although not 100%) in CI. Example:

https://circleci.com/gh/palantir/witchcraft-go-server/1022

Sample output:

```
=== RUN   TestNewInflightLimitMiddleware
--- FAIL: TestNewInflightLimitMiddleware (60.21s)
    ratelimit_test.go:101: 
        	Error Trace:	ratelimit_test.go:101
        	            				ratelimit_test.go:117
        	Error:      	Received unexpected error:
        	            	Post https://localhost:43351/example/post: context deadline exceeded
        	Test:       	TestNewInflightLimitMiddleware
    ratelimit_test.go:101: 
        	Error Trace:	ratelimit_test.go:101
        	            				ratelimit_test.go:112
        	            				asm_amd64.s:1333
        	Error:      	Received unexpected error:
        	            	Post https://localhost:43351/example/post: context deadline exceeded
        	Test:       	TestNewInflightLimitMiddleware
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x84cadd]

goroutine 221 [running]:
testing.tRunner.func1(0xc0003f6100)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x8dd720, 0xd74e20)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/palantir/witchcraft-go-server/integration.TestNewInflightLimitMiddleware(0xc0003f6100)
	/go/src/github.com/palantir/witchcraft-go-server/integration/ratelimit_test.go:118 +0x72d
testing.tRunner(0xc0003f6100, 0x991ea0)
	/usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x35c
FAIL	github.com/palantir/witchcraft-go-server/integration                                        	61.522s
```

--- a/changelog/@unreleased/pr-145.v2.yml
+++ b/changelog/@unreleased/pr-145.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: wait for requests before testing rate limit
+  links:
+  - https://github.com/palantir/witchcraft-go-server/pull/145

--- a/changelog/@unreleased/pr-145.v2.yml
+++ b/changelog/@unreleased/pr-145.v2.yml
@@ -1,5 +1,5 @@
 type: fix
 fix:
-  description: wait for requests before testing rate limit
+  description: Fix flaky test by waiting for requests before testing rate limit
   links:
   - https://github.com/palantir/witchcraft-go-server/pull/145

--- a/integration/ratelimit_test.go
+++ b/integration/ratelimit_test.go
@@ -49,7 +49,7 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 	wait, closeWait := context.WithCancel(context.Background())
 	defer closeWait()
 
-	totalPostRequests := 4
+	const totalPostRequests = 4
 	reqChan := make(chan struct{}, totalPostRequests)
 	initFn := func(ctx context.Context, info witchcraft.InitInfo) (cleanup func(), rErr error) {
 		info.Router.RootRouter().AddRouteHandlerMiddleware(limiter)
@@ -79,7 +79,7 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 
 	client := testServerClient()
 
-	testTimeout := time.Minute
+	const testTimeout = time.Minute
 	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
 	defer cancel()
 

--- a/integration/ratelimit_test.go
+++ b/integration/ratelimit_test.go
@@ -23,6 +23,7 @@ import (
 	"time"
 
 	"github.com/palantir/pkg/httpserver"
+	werror "github.com/palantir/witchcraft-go-error"
 	"github.com/palantir/witchcraft-go-server/conjure/witchcraft/api/health"
 	"github.com/palantir/witchcraft-go-server/status/reporter"
 	"github.com/palantir/witchcraft-go-server/witchcraft"
@@ -48,6 +49,8 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 	wait, closeWait := context.WithCancel(context.Background())
 	defer closeWait()
 
+	totalPostRequests := 4
+	reqChan := make(chan struct{}, totalPostRequests)
 	initFn := func(ctx context.Context, info witchcraft.InitInfo) (cleanup func(), rErr error) {
 		info.Router.RootRouter().AddRouteHandlerMiddleware(limiter)
 		if err := info.Router.Get("/get", http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
@@ -56,6 +59,7 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 			return nil, err
 		}
 		if err := info.Router.Post("/post", http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			reqChan <- struct{}{}
 			<-wait.Done()
 			rw.WriteHeader(http.StatusOK)
 		})); err != nil {
@@ -75,7 +79,8 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 
 	client := testServerClient()
 
-	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	testTimeout := time.Minute
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
 	defer cancel()
 
 	doGet := func() *http.Response {
@@ -111,7 +116,9 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 	requireHealthy("expected healthy before the second request")
 	go func() { resp2c <- doPost() }()
 
-	time.Sleep(time.Millisecond) // let things settle
+	// wait until both of the above requests have made it through
+	err = waitForRequests(reqChan, 2, testTimeout)
+	require.NoError(t, err)
 	requireHealthy("expected healthy before the third request")
 
 	resp3 := doPost()
@@ -138,5 +145,23 @@ func TestNewInflightLimitMiddleware(t *testing.T) {
 	case err := <-serverErr:
 		require.NoError(t, err)
 	default:
+	}
+}
+
+func waitForRequests(reqChan <-chan struct{}, expected int, timeout time.Duration) error {
+	t := time.After(timeout)
+	var current int
+	for {
+		select {
+		case <-reqChan:
+			current++
+			if current == expected {
+				return nil
+			}
+		case <-t:
+			return werror.Error("timed out waiting for expected number of requests",
+				werror.SafeParam("current", current),
+				werror.SafeParam("expected", expected))
+		}
 	}
 }

